include/bluetooth/mesh/mgmt.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2019 Tobias Svehagen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_MESH_MGMT_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_MESH_MGMT_H_
+
+#if defined(CONFIG_BT_MESH_NETWORK_MGMT)
+#define NODE_COUNT    CONFIG_BT_MESH_MGMT_NODE_COUNT
+#define SUBNET_COUNT  CONFIG_BT_MESH_MGMT_SUBNET_COUNT
+#define APP_KEY_COUNT CONFIG_BT_MESH_MGMT_APP_KEY_COUNT
+#else
+#define NODE_COUNT    0
+#define SUBNET_COUNT  0
+#define APP_KEY_COUNT 0
+#endif
+
+struct bt_mesh_mgmt_node {
+	u16_t addr;
+	u16_t net_idx;
+	u8_t  dev_key[16];
+	u8_t  num_elem;
+};
+
+struct bt_mesh_mgmt_subnet {
+	u16_t net_idx;
+
+	bool kr_flag;
+	u8_t kr_phase;
+
+	struct {
+		u8_t net_key[16];
+	} keys[2];
+};
+
+struct bt_mesh_mgmt_app_key {
+	u16_t net_idx;
+	u16_t app_idx;
+	struct {
+		u8_t id;
+		u8_t key[16];
+	} keys[2];
+};
+
+enum {
+	BT_MESH_DB_VALID,
+	BT_MESH_DB_NET_PENDING,
+	BT_MESH_DB_KEYS_PENDING,
+	BT_MESH_DB_NODES_PENDING,
+
+
+	/* Don't touch - intentionally last */
+	BT_MESH_DB_FLAG_COUNT,
+};
+
+struct bt_mesh_mgmt_net {
+	u32_t iv_index;
+
+	ATOMIC_DEFINE(flags, BT_MESH_DB_FLAG_COUNT);
+
+	struct bt_mesh_mgmt_node nodes[NODE_COUNT];
+	struct bt_mesh_mgmt_subnet subnets[SUBNET_COUNT];
+	struct bt_mesh_mgmt_app_key app_keys[APP_KEY_COUNT];
+};
+
+extern struct bt_mesh_mgmt_net bt_mesh_network;
+
+struct bt_mesh_mgmt_subnet *bt_mesh_mgmt_subnet_alloc(u16_t net_idx);
```

Comment:
Each subnet has two keys to handle key refresh. In `settings.c:db_subnet_set()` it is allocated and then the keys and key refresh is set. That could be done when overriding as well? Otherwise, maybe we must provide two keys in `bt_mesh_mgmt_subnet_alloc()`?

Reply:
While on the subject, have you thought about how the flags for the subnet will be kept updated? I.e. during key refresh and IV update, that information is received and stored by a node, `kr_flag` in `struct bt_mesh_subnet` and `BT_MESH_IVU_IN_PROGRESS` in `struct bt_mesh`. Should we also update the subnet in the database when updating the subnet for the local node?

The thing is that the `flags` field is part of the `prov_data` during provisioning. If a provisioner is supposed to be able to provision to networks that it is not a part of, how would it know things like if an IV update is in progress. :thinking:

Reply:
> Each subnet has two keys to handle key refresh

That's a good point, I hadn't thought about that. I think this can happen in a different function call. If the user really wants to establish a main key and a key refresh key, they call this function first, then the KR-function to register the KR key, perhaps?

> Should we also update the subnet in the database when updating the subnet for the local node?

Another good point :smile: I think the key refresh flag should definitely be part of the database, since that is controlled by the configurator device. I'm not so sure about IVU and IV index though. Those states are owned and controlled by the network as a collective. Duplicating them in the db would require infrastructure for listening for changes, and there's no scenario where the db value should be different from the local value, right? 

On a related note, the configurator also needs to know which devices have received the new key during key refresh. Perhaps we could limit the key refresh to one key at the time, and add a KR phase state to `bt_mesh_db_node` as well? I don't think this necessarily needs to be stored persistently, since we can just start over if we lose power, but we have to keep track of it somehow.